docs/design/validation.md

```diff
@@ -0,0 +1,1989 @@
+type ValidateResourceConfigRequest struct {
+    Config   tfsdk.Config
+    TypeName string
+}
+
+// ValidateResourceConfigResponse contains request information for the ValidateResourceConfig RPC.
+type ValidateResourceConfigResponse struct {
+    Diagnostics []*tfprotov6.Diagnostic
+}
+
+// ResourceConfigValidator describes a reusable Resource configuration validation function.
+type ResourceConfigValidator interface {
+    Description(context.Context) string
+    MarkdownDescription(context.Context) string
+    Validate(context.Context, ValidateResourceConfigRequest, *ValidateResourceConfigResponse)
+}
+
+// ResourceConfigValidatorWithProvider is an interface type for declaring configuration validation that requires a provider instance.
+type ResourceConfigValidatorWithProvider interface {
+    ResourceConfigValidator
+    ValidateWithProvider(context.Context, tfsdk.Provider, ValidateResourceConfigRequest, *ValidateResourceConfigResponse)
+}
+
+// ResourceWithConfigValidators is an interface type that extends Resource to include declarative validations.
+type ResourceWithConfigValidators interface {
+    Resource
+    ConfigValidators(context.Context) []ResourceConfigValidator
+}
+
+// ResourceWithValidateConfig is an interface type that extends Resource to include imperative validations.
+type ResourceWithValidateConfig interface {
+    Resource
+    ValidateConfig(context.Context, ValidateResourceConfigRequest, *ValidateResourceConfigResponse)
+}
+```
+
+Example provider code:
+
+```go
+func (r *customResource) ConfigValidators(ctx context.Context) ResourceConfigValidators {
+    return ResourceConfigValidators{
+        ConflictingAttributes(
+            tftypes.NewAttributePath().AttributeName("first_attribute"),
+            tftypes.NewAttributePath().AttributeName("second_attribute"),
+        ),
+    }
+}
+```
+
+### Attribute Level Example Implementation
+
+Example framework code:
+
+```go
+type ValidateAttributeRequest struct {
+    // AttributePath contains the path of the attribute.
+    AttributePath tftypes.AttributePath
+
+    // AttributeConfig contains the value of the attribute.
+    AttributeConfig attr.Value
+
+    // Config contains the entire configuration of the data source, provider, or resource.
+    Config tfsdk.Config
+}
+
+type ValidateAttributeResponse struct {
+    Diagnostics []*tfprotov6.Diagnostic
+}
+
+// AttributeValidator describes attribute validation functionality.
+type AttributeValidator interface {
+    Description(context.Context) string
+    MarkdownDescription(context.Context) string
+    Validate(context.Context, ValidateAttributeRequest, *ValidateAttributeResponse)
+}
+
+// Existing schema.Attribute struct type
+type Attribute struct {
+    // ...
+    Validators []AttributeValidator
+}
+```
+
+Example validation function code:
+
+```go
+type stringLengthBetweenValidator struct {
+    StringValueValidator
```

Comment:
What is this type? I don't think, based on the code you've supplied, this is necessary.

Reply:
Sorry, older code based on the strongly typed sketches, will update next commit.